dev/usb/class/cdcserial/cdcserial.c

```diff
@@ -0,0 +1,331 @@
+    0x00,   /* bmCapabilities: D0+D1 */
+    0x00,   /* bDataInterface: 1 */
+
+    /*ACM Functional Descriptor*/
+    0x04,   /* bFunctionLength */
+    0x24,   /* bDescriptorType: CS_INTERFACE */
+    0x02,   /* bDescriptorSubtype: Abstract Control Management desc */
+    0x02,   /* bmCapabilities */
+
+    /*Union Functional Descriptor*/
+    0x05,   /* bFunctionLength */
+    0x24,   /* bDescriptorType: CS_INTERFACE */
+    0x06,   /* bDescriptorSubtype: Union func desc */
+    0x00,   /* bMasterInterface: Communication class interface */
+    0x01,   /* bSlaveInterface0: Data Class Interface */
+
+    /* endpoint 1 IN */
+    0x07,           /* length */
+    ENDPOINT,       /* type */
+    0x82,           /* address: 1 IN */
+    0x03,           /* type: bulk */
+    W(8),           /* max packet size: 64 */
+    0xFF,           /* interval */
+};
+
+static const uint8_t data_if_descriptor[] = {
+    /*Data class interface descriptor*/
+    0x09,       /* bLength: Endpoint Descriptor size */
+    INTERFACE,  /* bDescriptorType: */
+    0x01,       /* bInterfaceNumber: Number of Interface */
+    0x00,       /* bAlternateSetting: Alternate setting */
+    0x02,       /* bNumEndpoints: Two endpoints used */
+    0x0A,       /* bInterfaceClass: CDC */
+    0x00,       /* bInterfaceSubClass: */
+    0x00,       /* bInterfaceProtocol: */
+    0x00,       /* iInterface: */
+
+    /*Endpoint OUT Descriptor*/
+    0x07,      /* bLength: Endpoint Descriptor size */
+    ENDPOINT,  /* bDescriptorType: Endpoint */
+    0x01,      /* bEndpointAddress */
+    0x02,      /* bmAttributes: Bulk */
+    0x40,      /* wMaxPacketSize: */
+    0x00,
+    0x00,      /* bInterval: ignore for Bulk transfer */
+
+    /*Endpoint IN Descriptor*/
+    0x07,      /* bLength: Endpoint Descriptor size */
+    ENDPOINT,  /* bDescriptorType: Endpoint */
+    0x81,      /* bEndpointAddress */
+    0x02,      /* bmAttributes: Bulk */
+    0x40,      /* wMaxPacketSize: */
+    0x00,
+    0x00       /* bInterval */
+};
+
+static const uchar langid[] = { 0x04, 0x03, 0x09, 0x04 };
+
+usb_config config = {
+    .lowspeed = {
+        .device = USB_DESC_STATIC(dev_descr),
+        .device_qual = USB_DESC_STATIC(devqual_descr),
+        .config = USB_DESC_STATIC(cfg_descr),
+    },
+    .highspeed = {
+        .device = USB_DESC_STATIC(dev_descr),
+        .device_qual = USB_DESC_STATIC(devqual_descr),
+        .config = USB_DESC_STATIC(cfg_descr),
+    },
+
+    .langid = USB_DESC_STATIC(langid),
+};
+
+static status_t usb_register_cb(
+    void *cookie,
+    usb_callback_op_t op,
+    const union usb_callback_args *args
+)
+{
+    if (op == USB_CB_ONLINE) {
+        usbc_setup_endpoint(1, USB_IN, 0x40, USB_BULK);
+        usbc_setup_endpoint(1, USB_OUT, 0x40, USB_BULK);
+        usbc_setup_endpoint(2, USB_IN, 0x40, USB_INTR);
+
+        usb_online = true;
+    }
+    return NO_ERROR;
+}
+
+static status_t usb_setup_cb(
```

Comment:
as pointed out below, you dont really need a separate callback here, this can be funneled through the regular CB as another opcode. Returns from all the register callbacks gets a little more complicated in that case, but if you call all of the registered CB's and then check if any of them handled the setup message it should work.